File: 67. Add Binary/add-binary.c
// 67. Add Binary (3/21/2022)
// Runtime: 2 ms (46.57%) Memory: 5.61 MB (78.02%) 

#define MAX(a, b) ((a > b) ? a : b)

char* addBinary (char* a, char* b) {
    int len_a = strlen(a), len_b = strlen(b);
    char* result = (char*)calloc(MAX(len_a, len_b) + 2, sizeof(char));
    
    int carry = 0;
    int idx_a = strlen(a) - 1, idx_b = strlen(b) - 1;
    for (int i = MAX(len_a, len_b); i >= 0; i--) {
        int bit = carry;
        if (idx_a >= 0) bit += a[idx_a--] - '0';
        if (idx_b >= 0) bit += b[idx_b--] - '0';
        carry = bit >> 1;
        result[i] = (bit % 2) + '0';
    }
    
    return result + (result[0] == '0');
}

File: 67. Add Binary/add-binary.ts
// 67. Add Binary (7/18/54188)
// Runtime: 76 ms (78.57%) Memory: 45.57 MB (26.89%) 

function addBinary(a: string, b: string): string {
    let big_a = BigInt(0), big_b = BigInt(0);

    for (let i = 0; i < a.length; i++)
        big_a += BigInt(a[i]) << BigInt(a.length - i - 1);
    for (let i = 0; i < b.length; i++)
        big_b += BigInt(b[i]) << BigInt(b.length - i - 1);

    return (big_a + big_b).toString(2);
};
